Text BTLxProcessing

obucklin opened the pull request:
New `Text` BTLx Processing.
Still needs unit tests. 
Still needs a workflow for adding the processing based on element attributes, placement, etc. 
Still needs a way to visualize geometry.


- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

obucklin commented:
Hey Panos, lets work on this together to see if we can polish it up.

chenkasirer commented on src/compas_timber/fabrication/basic_characters.json:
what's this file for?

obucklin replied:
This contains the polylines to draw the text characters

chenkasirer replied:
it rather large. maybe try to save it as not prettified (not indentations) or even take a look at: https://compas.dev/compas/latest/api/generated/compas.data.json_dumpz.html

chenkasirer commented on src/compas_timber/ghpython/components/CT_Label_Beams/code.py:
```diff
@@ -0,0 +1,35 @@
+from compas_timber.fabrication import Text
+
+
+class MyComponent(component):
```
please give meaningful name

chenkasirer commented on src/compas_timber/ghpython/components/CT_Label_Beams/code.py:
```diff
@@ -0,0 +1,35 @@
+class MyComponent(component):
+    def RunScript(self, model, base_string=None, char_to_replace=None, attributes=None, text_size=None):
+        scene = Scene()
+        self.clear_labels(model)
+        for element in model.beams:
+            for feature in element.features:
+                if getattr(feature, "is_label", None):
+                    element.remove_features(feature)
```
aren't these two doing the same?

obucklin replied:
see comment above

chenkasirer commented on src/compas_timber/ghpython/components/CT_Label_Beams/code.py:
```diff
@@ -0,0 +1,35 @@
+        self.clear_labels(model)
+        for element in model.beams:
+            for feature in element.features:
+                if getattr(feature, "is_label", None):
```
I find it better to check `feature.PROCESSING_NAME == "Text"` here then to extend the public interface of just that one process with `is_label`

obucklin replied:
The idea was to leave `Text` processings that were generated elsewhere. I tried to do this with a persistent list in the component, but I couldn't figure that out. The problem is that this modifies the elements in the original model, so if any parameters change, you end up with multiple labels. I see a few options:
-Copy the model in this component and add the labels to the new elements.
-give the elements a `label` attribute which is a single `Text` processing. this would be updated by the `LabelBeams` component.
-give the processings a `is_label` boolean, allowing us to identify and remove them.

@chenkasirer what do you think? any other options?

obucklin replied:
> The idea was to leave `Text` processings that were generated elsewhere. I tried to do this with a persistent list in the component, but I couldn't figure that out. The problem is that this modifies the elements in the original model, so if any parameters change, you end up with multiple labels. I see a few options: -Copy the model in this component and add the labels to the new elements. -give the elements a `label` attribute which is a single `Text` processing. this would be updated by the `LabelBeams` component. -give the processings a `is_label` boolean, allowing us to identify and remove them.
> 
> @chenkasirer what do you think? any other options?

@chenkasirer you have any thoughts about how to do this?

chenkasirer replied:
mmh my I like the idea of a "LabelBeams" component where you tell it which attribute of of the beam you want and it takes and uses it as the text. If the component spits out a `BTLxFromGeometryDefinition`, then should fit nicely within the current system.. or am I missing something? 

Actually setting the attribute could be done as part of the BeamFromCenterline componet or outside the scope of this (AKA should already be set like e.g. the key)

obucklin replied:
The idea was that this label also responds to attributes generated at the `model` phase and would include attributes like `group`, assembly order, etc. I guess technically that stuff is already in the element component, but in terms of workflow it makes sense to have this post-process the `model.elements`.

chenkasirer commented on src/compas_timber/ghpython/components/CT_Label_Beams/metadata.json:
```diff
@@ -0,0 +1,54 @@
+{
```
copy paste accident? ;)

obucklin commented:
Hey @papachap @chenkasirer,  I reworked this to have a better GH workflow and to hopefully permit more robust expansion later. I essentially did the following:

Added a `Label` class in `compas_timber.planning`. This allows us to define the content of the Label (the text, element attributes, etc) separately, which we can then pass to different label types, like a `Text` `BTLxProcessing`, but also to normal printed labels or stickers. This could be expanded to make human and machine-readable labels. 

`Label` also has a function to find the biggest free gap on the beam and to place the `Text` processing there. 

There is a `DeferredLabel` which allows us to pass the label to the model GH component.  This uses the same `feature_from_elements` method to get the `BTLxProcessing` allowing us to use it in the same way as the other feature defs.

I will work on some unit tests for the `Text` and `Label`, but if you have time, it would be great to get some feedback. have a nice weekend!

obucklin commented:
added unit tests. ready for review

chenkasirer commented on src/compas_timber/fabrication/text.py:
```diff
@@ -0,0 +1,285 @@
+        super(TextParams, self).__init__(instance)
+
+    def as_dict(self):
+        result = super(TextParams, self).as_dict()
```
here also, align with `main`

chenkasirer commented on src/compas_timber/fabrication/text.py:
```diff
@@ -0,0 +1,285 @@
+        """
+        return geometry
+
+    def draw_string_on_element(self, element):
```
I understand we're not doing this at the moment but we might want to in the future, right?
could you add a docstring so it's clearer what this does?

chenkasirer commented on src/compas_timber/planning/label.py:
```diff
@@ -0,0 +1,157 @@
+
+        return cls(element, string_out)
+
+    def engrave_on_beam(self, beam, text_height=None, ref_side_index=0):
```
should this maybe be an alternative constructor of the `Text`processing? (e.g. `Text.from_element_and_text`)

obucklin replied:
I see this as essentially like the `Joint.add_features()` method. I could add an alternate constructor to `Text` but I would keep this method here BC I think we want the option to call `Label.print_label` or `Label.mark_element` (yes some timber CNCs have marker tools)

chenkasirer commented on src/compas_timber/planning/label.py:
```diff
@@ -0,0 +1,157 @@
+        return "{}_{}_{}".format(self.element.__class__.__name__, self.element.key, self.element.attributes.get("category") if self.element.attributes.get("category") else "")
+
+    @classmethod
+    def from_element(cls, element, attributes=[], base_string="", char_to_replace=None):
```
immutable objects (`[]`) should not be set as default value as these are only evaluated once

chenkasirer commented on src/compas_timber/planning/__init__.py:
```diff
@@ -18,4 +20,6 @@
     "Step",
     "SimpleSequenceGenerator",
     "Text3d",
+    "Label",
+    "DeferredLabel",
```
dunno if that's the based place for these, also the BuildingPlan stuff might end up moving to compas XR.
how about the `workflow.py` module? is that not a better fit?

obucklin replied:
I don't have a real opinion on this, except that workflow is maybe becoming a bit of a catchall. I would leave this as its own module. Label is perhaps most relevant for fabrication, or maybe we need to consider an `assembly` package (for assembly processes).

chenkasirer commented on src/compas_timber/planning/label.py:
```diff
@@ -0,0 +1,157 @@
+        self.user_text = text
+
+    @property
+    def __data__(self):
```
we only need the data interface if this needs to be serialized. since it doesn't end on any beam maybe it doesn't. 
if it does, then it should inherit from `Data`

obucklin replied:
I'm still not sure on when we need to be able to serialize, and how to best do it. Shouldn't every compas object be serializable? When should a class inherit from `object` vs `Data`?

chenkasirer commented on src/compas_timber/planning/label.py:
```diff
@@ -0,0 +1,157 @@
+        return "DeferredLabel: {}".format(self.base_string)
+
+    @property
+    def __data__(self):
```
same here, not sure if this needs `Data`

obucklin replied:
see my above response

obucklin commented:
@chenkasirer not ready for review, but I have some questions and comments of my own.
